### PR TITLE
Refactor TransactionPool a bit

### PR DESCRIPTION
**Pull request by AndrejMitrovic**

### CODE REVIEW

**Comment by Geod24** [comment]
A bit??

**Comment by AndrejMitrovic** [comment]
Is there a way to wrap a range into another range and make the dtor @safe/@trusted?

It seems the `ResultRange` from the sqlite bindings has an unsafe destructor.

**Comment by AndrejMitrovic** [comment]
Added another commit. `Use the tx pool range in its constructor`

File: source/agora/common/TransactionPool.d [hunk 1]
```diff
@@ -193,32 +194,17 @@ public class TransactionPool
...
-            dg = the delegate to call
-
         Returns:
-            the status code of the delegate, or zero
+            an input range of (hash, transaction) tuples
 
     ***************************************************************************/
 
-    public int opApply (scope int delegate(Hash hash, Transaction tx) dg) @trusted
+    public auto opSlice () @trusted
```
Comment (Geod24) [inline source/agora/common/TransactionPool.d]: If the resulting range is not safe, isn't it better to not marked this as `@trusted` ?

File: source/agora/node/Ledger.d [hunk 2]
```diff
@@ -283,19 +283,19 @@ public class Ledger
 
         this.enroll_man.getUnregistered(data.enrolls);
 
-        foreach (hash, tx; this.pool)
-        {
-            if (auto reason = tx.isInvalidReason(utxo_finder, next_height))
-                log.trace("Rejected invalid ('{}') tx: {}", reason, tx);
-            else
-                data.tx_set.put(tx);
+        () @trusted {
```
Comment (Geod24) [inline source/agora/node/Ledger.d]: I'm not sure that's an enhancement to be fair...

**Review by Geod24 — COMMENTED** [review]
Not so convince about the range change. In spirit it's nice, but the result is, the calling code looks way worse.

File: source/agora/common/TransactionPool.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/common/TransactionPool.d]: Well the method itself is trusted, but the returned type is not. I wish I could do `@trusted(auto)` hah.

**Comment by AndrejMitrovic** [comment]
> Not so convince about the range change. In spirit it's nice, but the result is, the calling code looks way worse.

Yep it's because of the dtor.

Actually I wanted to try to do copy ellision. Right now the opApply copies hashes / txes.

**Comment by AndrejMitrovic** [comment]
I'll revert the range change for now.

**Review by MichaelKim20 — APPROVED** [review]
LGTM

**Comment by MichaelKim20** [comment]
I changed the order of the source code, rearranged it, and looked at it one by one. 
LGTM

**Review by Geod24 — APPROVED** [review]

**Pull request merged**